Compiler error when calling closure incorrectly using fn_traits feature

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```Rust
#![feature(fn_traits, unboxed_closures, tuple_trait)]

use std::marker::Tuple;
use std::default::Default;

fn wrap<P: Tuple + Default, T>(
    func: impl Fn<P, Output = T>
) {
    let x: P = Default::default();
    // Should be: `func.call(x);`
    func(x);
}

fn foo() { }

fn main() {
    wrap(foo);
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.68.0-nightly (0468a00ae 2022-12-17)
binary: rustc
commit-hash: 0468a00ae3fd6ef1a6a0f9eaf637d7aa9e604acc
commit-date: 2022-12-17
host: x86_64-pc-windows-msvc
release: 1.68.0-nightly
LLVM version: 15.0.6
```

### Error output

```
error: internal compiler error: no errors encountered even though `delay_span_bug` issued

error: internal compiler error[E0059]: cannot use call notation; the first type parameter for the function trait is neither a tuple nor unit
  --> src\main.rs:11:5
   |
11 |     func(x);
   |     ^^^^^^^

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler\rustc_hir_typeck\src\fn_ctxt\_impl.rs:685:23

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler\rustc_hir_typeck\src\coercion.rs:174:49

error: internal compiler error: broken MIR in DefId(0:5 ~ triage[3bb5]::wrap) (Terminator { source_info: SourceInfo { span: src\main.rs:11:5: 11:12 (#0), scope: scope[1] }, kind: _3 = <impl Fn<P, Output = T> as Fn<P>>::call(move _4, move _5) -> [return: bb2, unwind: bb8] }): bad arg #1 (P <- (P,)): NoSolution
  --> src\main.rs:11:5
   |
11 |     func(x);
   |     ^^^^^^^
   |
   = note: delayed at compiler\rustc_borrowck\src\type_check\mod.rs:1580:17
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'Box<dyn Any>', compiler\rustc_errors\src\lib.rs:1610:13
stack backtrace:
   0:     0x7ffff9959de2 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hea3121a92230cef2
   1:     0x7ffff9995d0b - core::fmt::write::h0bc3f53e8dca163a
   2:     0x7ffff994c8fa - <std::io::IoSlice as core::fmt::Debug>::fmt::h7890ee78f4472cf6
   3:     0x7ffff9959b2b - std::sys::common::alloc::realloc_fallback::h2905b67a6dea1c71
   4:     0x7ffff995d459 - std::panicking::default_hook::h111ce7f1b363c3ab
   5:     0x7ffff995d0db - std::panicking::default_hook::h111ce7f1b363c3ab
   6:     0x7ffff28c4deb - rustc_driver[5e4fd5aa1121bb7c]::describe_lints
   7:     0x7ffff995ddc0 - std::panicking::rust_panic_with_hook::hfc6ef76eb93085aa
   8:     0x7ffff29054a3 - <rustc_errors[2f3e2ae2073fef46]::ErrorGuaranteed as core[c581a5698e0001c6]::fmt::Debug>::fmt
   9:     0x7ffff2903959 - <rustc_errors[2f3e2ae2073fef46]::ErrorGuaranteed as core[c581a5698e0001c6]::fmt::Debug>::fmt
  10:     0x7ffff28e1e29 - <rustc_errors[2f3e2ae2073fef46]::json::JsonEmitter as rustc_errors[2f3e2ae2073fef46]::emitter::Emitter>::should_show_explain
  11:     0x7ffff2900249 - <rustc_span[25fa311e2ce9dc22]::symbol::MacroRulesNormalizedIdent as rustc_errors[2f3e2ae2073fef46]::diagnostic::IntoDiagnosticArg>::into_diagnostic_arg
  12:     0x7fffefaa5b7d - <rustc_errors[2f3e2ae2073fef46]::HandlerInner>::emit_diagnostic
  13:     0x7fffefc4116d - rustc_interface[e09bab31b5025fb2]::passes::prepare_outputs
  14:     0x7fffefc5fe08 - <rustc_interface[e09bab31b5025fb2]::queries::Queries>::ongoing_codegen
  15:     0x7fffefa5434c - rustc_driver[5e4fd5aa1121bb7c]::args::arg_expand_all
  16:     0x7fffefa44983 - <rustc_span[25fa311e2ce9dc22]::DebuggerVisualizerFile>::new
  17:     0x7fffefa3a24d - <rustc_span[25fa311e2ce9dc22]::DebuggerVisualizerFile>::new
  18:     0x7ffff996fe9c - std::sys::windows::thread::Thread::new::h4025a5964e95cbad
  19:     0x7ff84f7e7c24 - BaseThreadInitThunk
  20:     0x7ff85162d721 - RtlUserThreadStart
```

</p>
</details>

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
@@ -214,7 +214,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         "cannot use call notation; the first type parameter \
                          for the function trait is neither a tuple nor unit"
                     )
-                    .delay_as_bug();
+                    .emit();
                     (self.err_args(provided_args.len()), None)
                 }
             }

--- a/src/test/ui/unboxed-closures/non-tupled-call.rs
+++ b/src/test/ui/unboxed-closures/non-tupled-call.rs
@@ -1,0 +1,17 @@
+#![feature(fn_traits, unboxed_closures, tuple_trait)]
+
+use std::default::Default;
+use std::marker::Tuple;
+
+fn wrap<P: Tuple + Default, T>(func: impl Fn<P, Output = T>) {
+    let x: P = Default::default();
+    // Should be: `func.call(x);`
+    func(x);
+    //~^ ERROR cannot use call notation; the first type parameter for the function trait is neither a tuple nor unit
+}
+
+fn foo() {}
+
+fn main() {
+    wrap(foo);
+}

--- a/src/test/ui/unboxed-closures/non-tupled-call.stderr
+++ b/src/test/ui/unboxed-closures/non-tupled-call.stderr
@@ -1,0 +1,9 @@
+error[E0059]: cannot use call notation; the first type parameter for the function trait is neither a tuple nor unit
+  --> $DIR/non-tupled-call.rs:9:5
+   |
+LL |     func(x);
+   |     ^^^^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0059`.

